main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -130,70 +131,42 @@ object Inliner {
         case _ => LiftedAst.Expression.ApplyAtomic(op, es, tpe, purity, loc)
       }
 
-    case OccurrenceAst.Expression.ApplyClo(exp, args, tpe, purity, loc) =>
+    case OccurrenceAst.Expression.ApplyClo(exp, exps, ct, tpe, purity, loc) =>
       val e = visitExp(exp, subst0)
-      val as = args.map(visitExp(_, subst0))
+      val es = exps.map(visitExp(_, subst0))
       e match {
         case LiftedAst.Expression.ApplyAtomic(AtomicOp.Closure(sym), closureArgs, _, _, _) =>
           val def1 = root.defs.apply(sym)
           // If `def1` is a single non-self call or
           // it is trivial
           // then inline the body of `def1`
           if (canInlineDef(def1)) {
-            val e1 = rewriteTailCalls(def1.exp)
+            val e1 = ct match {
+              case CallType.TailCall => def1.exp
+              case CallType.NonTailCall => rewriteTailCalls(def1.exp)
+            }
             // Map for substituting formal parameters of a function with the closureArgs currently in scope
-            bindFormals(e1, (def1.cparams ++ def1.fparams).map(_.sym), closureArgs ++ as, Map.empty)
-          } else {
-            LiftedAst.Expression.ApplyClo(e, as, tpe, purity, loc)
-          }
-        case _ =>
-          val as = args.map(visitExp(_, subst0))
-          LiftedAst.Expression.ApplyClo(e, as, tpe, purity, loc)
-      }
-
-    case OccurrenceAst.Expression.ApplyDef(sym, args, tpe, purity, loc) =>
-      val as = args.map(visitExp(_, subst0))
```

Comment:
Have you preserved the meaning of this code?

Reply:
Yes, the only difference is that I've renamed `args` to `exps` and merged `ApplyDef` and `ApplyDefTail` cases. The only difference those cases had were whether to call `rewriteTailCalls(def1.exp)` or to use `def1.exp` for `indFormals`. This difference is handled by

```scala
val e1 = ct match {
          case CallType.TailCall => def1.exp
          case CallType.NonTailCall => rewriteTailCalls(def1.exp)
        }
bindFormals(e1, (def1.cparams ++ def1.fparams).map(_.sym), es, Map.empty)
```

Reply:
Same thing for `ApplyClo`. I approached it very incrementally by first adding `ct` to `ApplyX` before merging `ApplyXTail` into anything and just preserving semantics with that addition. Then I rewrote each pattern match to match directly on the value of `ct` i.e., `ApplyDef(sym, exps, Ast.CallType.NonTailCall,...)` to ensure that semantics were stilled preserved. Finally, I looked at how to merge similar cases.

Reply:
In short, I'm fairly confident, it has the same meaning as before :)

Reply:
OK. I will take a much closer looker later then.